core/src/main/scala/kafka/log/UnifiedLog.scala

```diff
@@ -674,21 +682,29 @@ class UnifiedLog(@volatile var logStartOffset: Long,
   }
 
   /**
-   * Rename the directory of the local log
+   * Rename the directory of the local log. If the log's directory is being renamed for async deletion due to a
+   * StopReplica request, then the shouldReinitialize parameter should be set to false, otherwise it should be set to true.
    *
+   * @param name The new name that this log's directory is being renamed to
+   * @param shouldReinitialize Whether the log's metadata should be reinitialized after renaming
    * @throws KafkaStorageException if rename fails
    */
-  def renameDir(name: String): Unit = {
+  def renameDir(name: String, shouldReinitialize: Boolean): Unit = {
     lock synchronized {
       maybeHandleIOException(s"Error while renaming dir for $topicPartition in log dir ${dir.getParent}") {
         // Flush partitionMetadata file before initializing again
         maybeFlushMetadataFile()
```

Comment:
@junrao Could this line trigger an `IOException` while deleting the replica? I guess that it should be rare given that the topic ID is rarely updates but that seems possible to me. I am not an expect in this area so I might be wrong.

Reply:
@dajac : Yes, this line could trigger an IOException, but is rare since in the common case, once the topicId is created, it doesn't change.

Reply:
In this case, I think that we should update `testStopReplicaWithDeletePartitionAndExistingPartitionAndNewerLeaderEpochAndIOException` to actually cover this rare case to ensure that KafkaStorageException is handled correctly on the stop replica path. What do you think?

Reply:
Yes, it's probably useful to cover that. What do you think @gitlw ?

Reply:
Yeah, I agree it's useful to cover that case.
But I feel this is kind of orthogonal to the change in this PR. I can use a separate PR to address that issue. Are you ok with that?

Reply:
It is not orthogonal in my opinion. `testStopReplicaWithDeletePartitionAndExistingPartitionAndNewerLeaderEpochAndIOException` was meant to ensure `IOException` is handled correctly on the stop replica path. The issue is that the change made is this PR makes the trigger used obsolete.

It seems that we could just replace the following block in `testStopReplicaWithExistingPartition`:

```
if (throwIOException) {
  // Delete the underlying directory to trigger an KafkaStorageException
  val dir = partition.log.get.dir
  Utils.delete(dir)
  dir.createNewFile()
}
```

by the following one:

```
if (throwIOException) {
  // Replace underlying PartitionMetadataFile with a mock which thrown
  // a KafkaStorageException when maybeFlush is called.
  val mockPartitionMetadataFile = mock(classOf[PartitionMetadataFile])
  when(mockPartitionMetadataFile.maybeFlush()).thenThrow(new KafkaStorageException())
  partition.log.get.partitionMetadataFile = Some(mockPartitionMetadataFile)
}
```

and we are good to go. What do you think?

Reply:
@dajac I see what you mean. I've incorporated this change in the last commit. Thanks for this suggestion!